### PR TITLE
Blog App: Add devise

### PR DESCRIPTION
**Pull request by mahdinoori2000**
What is implemented in this PR:

- Remove `current_user` method in ApplicationController.
- [Install and setup devise gem](https://github.com/heartcombo/devise#getting-started).

- Can register a new user.
- User logs in with a combination of email and password.
- Hashed password should be stored in the database.
- Ask for confirmation of email.
- Can reset password.

- [Modify the views in devise](https://github.com/heartcombo/devise#configuring-views) for registration and login to match [the wireframes from the Sneak Peek](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/sneak_peek.md#:~:text=For%20this%20project%20you%20will%20have%20full%20freedom%20in%20terms%20of%20visual%20design%20but%20you%20will%20need%20to%20keep%20the%20following%20wireframes%3A) and your styling.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,8 +1,15 @@
 class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
```
Comment (mckent05) [inline app/models/user.rb]: - You have done well so far, but according to the project requirements, please make sure your database is confirmable by adding the `:confirmable` module to your user model

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,4 +1,5 @@
 class UsersController < ApplicationController
+  before_action :authenticate_user!
```
Comment (mckent05) [inline app/controllers/users_controller.rb]: - You have done well so far, but according to the project requirements, please ensure that a user is authenticated before accessing any of your controller actions. Please add this to your application_controller👍

File: app/models/user.rb [hunk 3]
```diff
@@ -1,8 +1,15 @@
...
   has_many :posts, foreign_key: 'author_id'
   has_many :likes, foreign_key: 'author_id'
   has_many :comments, foreign_key: 'author_id'
 
+  validates :email, presence: true, allow_blank: false
+  validates :encrypted_password, presence: true, allow_blank: false
```
Comment (mckent05) [inline app/models/user.rb]: **OPTIONAL** - Good work done so far, but you do not need to add these validations as they are already handled by devise. Please fix this 😁

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @mahdinoori2000, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍 
- Good job adding authentication to your app 👍 👍 


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/devise/registrations/new.html.erb
Comment (shahadat3669) [inline app/views/devise/registrations/new.html.erb:1]: [Optional] Well done! your project meets the previous reviewer's requirements. But still needs some changes.🔍 Let's enhance the user creation process by adding functionality to capture additional details such as `bio` and `photo`. Here are the steps to address this:

- [ ]  Please modify the user creation form to include additional input fields for `bio` and `photo`. Update the registration view to prompt users for these details during the account creation process.

- [ ] By default Device does not allow additional paramere. you have to configure those parameters to permit the additional attributes (`bio` and `photo`). Please follow the documentation in the official Devise GitHub repository, specifically the [Strong Parameters](https://github.com/heartcombo/devise?tab=readme-ov-file#strong-parameters) section, to set up additional parameters for the additional user attributes.

- [ ] Please Ensure that the registration controller is updated to handle the new parameters during user creation. Modify the controller actions to permit and save the new attributes.

Example code:
`app/controllers/application_controller.rb`
```
before_action :update_allowed_parameters, if: :devise_controller?
protected

 def update_allowed_parameters
    devise_parameter_sanitizer.permit(:sign_up) do |u|
      u.permit(:name, :surname, :photo, :bio, :email, :password, :password_confirmation)
    end
    devise_parameter_sanitizer.permit(:account_update) do |u|
      u.permit(:name, :surname, :photo, :bio, :email, :password, :current_password)
    end
  end

```

Implementing these changes will allow users to provide `bio` and `photo` details during account creation, enhancing the overall user profile functionality. Please test the updated registration process thoroughly to confirm that the new attributes are captured and stored correctly. Good luck! 🌟🚀

**Review by shahadat3669 — APPROVED** [review]
Hi @mahdinoori2000,  @MohammadYaser,

Your project is complete!  There is nothing else to say other than it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
✅ Email conformation working fine
✅ a user is authenticated before accessing any of controller actions
✅ everyone has contributed

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table  width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadat3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**